### PR TITLE
Batchnormgrad, squared difference, relu6grad optimized kernels

### PR DESCRIPTION
**Pull request by Jamather**

### CODE REVIEW

File: third_party/dml/DirectMLX.h [hunk 1]
```diff
@@ -1244,6 +1265,11 @@ namespace dml
         return detail::ElementWiseUnary<DML_OPERATOR_ELEMENT_WISE_SQRT, DML_ELEMENT_WISE_SQRT_OPERATOR_DESC>(input, scaleBias);
     }
 
+    inline Expression DifferenceSquare(Expression a, Expression b)
+    {
+        return detail::ElementWiseBinary<(DML_OPERATOR_TYPE)DML_OPERATOR_ELEMENT_WISE_DIFFERENCE_SQUARE, DML_ELEMENT_WISE_DIFFERENCE_SQUARE_OPERATOR_DESC>(a, b);
```
Comment (jstoecker) [inline third_party/dml/DirectMLX.h]:
```suggestion
        return detail::ElementWiseBinary<DML_OPERATOR_ELEMENT_WISE_DIFFERENCE_SQUARE, DML_ELEMENT_WISE_DIFFERENCE_SQUARE_OPERATOR_DESC>(a, b);
```

The cast isn't necessary. Presumably from when this was a preview enum.

File: third_party/dml/DirectMLX.h [hunk 2]
```diff
@@ -1074,6 +1074,27 @@ namespace dml
         return output;
     }
 
+    inline Expression ClipGrad(Expression input, Expression inputGradient, float min, float max)
```
Comment (jstoecker) [inline third_party/dml/DirectMLX.h:1077]: In general, changes to DirectMLX.h should go into the [copy in the DirectML repo first](https://github.com/microsoft/DirectML/blob/master/Libraries/DirectMLX.h). The order doesn't really matter, but just make sure to copy your changes over to that repo. At some point we'll want TF repo to consume the header using an external bazel repo instead of including a copy.

File: third_party/dml/DirectMLX.h [hunk 3]
```diff
@@ -1074,6 +1074,27 @@ namespace dml
...
+        TensorDesc outputGradientTensor(inputGradientTensor.dataType, inputGradientTensor.sizes, builder->GetTensorPolicy());
+
+        DML_ELEMENT_WISE_CLIP_GRAD_OPERATOR_DESC desc = {};
+        desc.InputTensor = inputTensor.AsPtr<DML_TENSOR_DESC>();
+        desc.OutputGradientTensor = outputGradientTensor.AsPtr<DML_TENSOR_DESC>();
+        desc.Min = min;
+        desc.Max = max;
+
+        detail::NodeOutput* const inputs[] = { input.Impl() };
+        detail::NodeID node = builder->CreateOperatorNode(DML_OPERATOR_ELEMENT_WISE_CLIP_GRAD, &desc, inputs);
```
Comment (jstoecker) [inline third_party/dml/DirectMLX.h:1093]: We might want to guard some of these helpers with DML_TARGET_VERSION; otherwise, using a newer copy of DirectMLX.h with an older DirectML.h will break. The D3D helper header didn't do this initially (AFAIK), but users sometimes [ran into issues](https://stackoverflow.com/questions/65294611/d3dx12-h-gives-a-bunch-of-errors).

At one point we discussed bundling DirectMLX.h inside the redistributable packages so this type of error wouldn't occur, but it won't help anyone targeting the inbox/SDK version. Pinging @adtsai for his thoughts on this.

File: third_party/dml/DirectMLX.h [hunk 4]
```diff
@@ -2644,6 +2670,48 @@ namespace dml
         return output;
     }
 
+    inline Expression BatchNormalizationGrad(
+        Expression inputX,
```
Comment (jstoecker) [inline third_party/dml/DirectMLX.h]: Why `inputX` and `inputYBackprop`? Why not just `input` and `inputGradient` to match the API struct? This is a little confusing.

**Review by jstoecker — CHANGES_REQUESTED** [review]

File: third_party/dml/DirectMLX.h (hunk 4 above)
Comment (adtsai) [reply third_party/dml/DirectMLX.h]: Yes, these need to match the public API.

File: third_party/dml/DirectMLX.h [hunk 5]
```diff
@@ -2644,6 +2670,48 @@ namespace dml
...
+        Expression inputYBackprop,
+        Expression mean,
+        Expression variance,
+        Expression scale,
+        float epsilon,
+        Expression &outputScaleGradient,
```
Comment (adtsai) [inline third_party/dml/DirectMLX.h]: For operators that return multiple tensors, you should return a struct rather than using (ref!) out params. See GRU or RandomGenerator for example.

File: third_party/dml/DirectMLX.h [hunk 6]
```diff
@@ -1074,6 +1074,27 @@ namespace dml
...
+    {
+        detail::GraphBuilder* builder = input.Impl()->GetGraphBuilder();
+
+        TensorDesc inputTensor = input.Impl()->GetOutputDesc();
+        TensorDesc inputGradientTensor = inputGradient.Impl()->GetOutputDesc();
+        TensorDesc outputGradientTensor(inputGradientTensor.dataType, inputGradientTensor.sizes, builder->GetTensorPolicy());
+
+        DML_ELEMENT_WISE_CLIP_GRAD_OPERATOR_DESC desc = {};
+        desc.InputTensor = inputTensor.AsPtr<DML_TENSOR_DESC>();
+        desc.OutputGradientTensor = outputGradientTensor.AsPtr<DML_TENSOR_DESC>();
```
Comment (adtsai) [inline third_party/dml/DirectMLX.h:1088]: Aren't you forgetting to set InputGradientTensor here?

File: tensorflow/core/kernels/dml_batch_norm_ops.cc [hunk 7]
```diff
@@ -776,32 +787,33 @@ class DmlFusedBatchNormGradKernel : public DmlKernel {
...
+          dml::Reduce(coef1, DML_REDUCE_FUNCTION_SUM, {0, 2, 3}) * coef0;
 
-    // offset_backprop = sum(y_backprop)
-    auto offset_backprop =
-        dml::Reduce(y_backprop, DML_REDUCE_FUNCTION_SUM, {0, 2, 3});
+      // offset_backprop = sum(y_backprop)
+      offset_backprop =
+          dml::Reduce(y_backprop, DML_REDUCE_FUNCTION_SUM, {0, 2, 3});
+    }
+    else
```
Comment (adtsai) [inline tensorflow/core/kernels/dml_batch_norm_ops.cc]: You need to turn on clang-format.

**Review by adtsai — CHANGES_REQUESTED** [review]
:clock1:

File: third_party/dml/DirectMLX.h (hunk 4 above)
Comment (Jamather) [reply third_party/dml/DirectMLX.h]: I wanted to refer to it as x on the TF side of things, but of course this is a DML header so that's no good here :p

**Review by jstoecker — APPROVED** [review]
Approving pending any changes Adrian may want regarding feature level macro guards.

File: third_party/dml/DirectMLX.h [hunk 8]
```diff
@@ -2644,6 +2671,54 @@ namespace dml
         return output;
     }
 
+    struct BatchNormalizationGradOutputs
+    {
+        Expression outputGradient;
```
Comment (adtsai) [inline third_party/dml/DirectMLX.h]: The convention for these output structs is to omit "output" from the field names. See `RandomGeneratorOutputs`, `MaxPoolingOutputs`, and `GRUOutputs` for example.

**Comment by adtsai** [comment]
> Approving pending any changes Adrian may want regarding feature level macro guards.

I won't block on that for this PR, but I'll add it to my to-do list to look into in the future - it'd be a worthy thing to add.

**Comment by Jamather** [comment]
sqdif and relu6 tests pass without a hitch. Batchnormgrad fails some tests with some elements having an absolute difference of >0.002 but <0.004

[       OK ] BatchNormalizationTest.testBatchNormGradGradConfig1
[ RUN      ] BatchNormalizationTest.testBatchNormGradGradConfig2
[       OK ] BatchNormalizationTest.testBatchNormGradGradConfig2
[ RUN      ] BatchNormalizationTest.testBatchNormGradGradConfig3
[       OK ] BatchNormalizationTest.testBatchNormGradGradConfig3
[ RUN      ] BatchNormalizationTest.testBatchNormGradGradConfig4
[       OK ] BatchNormalizationTest.testBatchNormGradGradConfig4
[ RUN      ] BatchNormalizationTest.testBatchNormGradShape1
[  FAILED  ] BatchNormalizationTest.testBatchNormGradShape1
[ RUN      ] BatchNormalizationTest.testBatchNormGradShape2
[  FAILED  ] BatchNormalizationTest.testBatchNormGradShape2
[ RUN      ] BatchNormalizationTest.testBatchNormGradShape3
[  FAILED  ] BatchNormalizationTest.testBatchNormGradShape3
[ RUN      ] BatchNormalizationTest.testBatchNormGradShape4
[  FAILED  ] BatchNormalizationTest.testBatchNormGradShape4
[ RUN      ] BatchNormalizationTest.testBatchNormGradShape5
[       OK ] BatchNormalizationTest.testBatchNormGradShape5
[ RUN      ] BatchNormalizationTest.testInferenceShape1
[       OK ] BatchNormalizationTest.testInferenceShape1
[ RUN      ] BatchNormalizationTest.testInferenceShape2
[       OK ] BatchNormalizationTest.testInferenceShape2
[ RUN      ] BatchNormalizationTest.testInferenceShape3
[       OK ] BatchNormalizationTest.testInferenceShape3
[ RUN      ] BatchNormalizationTest.testInferenceShape4
[  FAILED  ] BatchNormalizationTest.testInferenceShape4
[ RUN      ] BatchNormalizationTest.testInferenceShape5
[       OK ] BatchNormalizationTest.testInferenceShape5
[ RUN      ] BatchNormalizationTest.testTrainingShape1
[       OK ] BatchNormalizationTest.testTrainingShape1
[ RUN      ] BatchNormalizationTest.testTrainingShape2
[       OK ] BatchNormalizationTest.testTrainingShape2
[ RUN      ] BatchNormalizationTest.testTrainingShape3
[       OK ] BatchNormalizationTest.testTrainingShape3
[ RUN      ] BatchNormalizationTest.testTrainingShape4
[  FAILED  ] BatchNormalizationTest.testTrainingShape4
[ RUN      ] BatchNormalizationTest.testTrainingShape5
[  SKIPPED ] BatchNormalizationTest.testTrainingShape5

**Review by adtsai — APPROVED** [review]
:shipit:
